src/daft-schema/src/schema.rs

```diff
@@ -10,99 +9,128 @@ use common_display::{
-                        entry.key()
-                    )));
-                }
+    pub fn new(fields: Vec<Field>) -> Self {
+        let mut name_to_indices = HashMap::<String, Vec<usize>>::new();
+
+        for (i, f) in fields.iter().enumerate() {
+            if let Some(indices) = name_to_indices.get_mut(&f.name) {
+                indices.push(i);
+            } else {
+                name_to_indices.insert(f.name.clone(), vec![i]);
             }
         }
 
-        Ok(Self { fields: map })
+        Self {
+            fields,
+            name_to_indices,
+        }
+    }
+
+    pub fn empty() -> Self {
+        Self::new(vec![])
     }
 
     pub fn to_struct(&self) -> DataType {
-        let fields = self.fields.values().cloned().collect();
-        DataType::Struct(fields)
+        DataType::Struct(self.fields.clone())
     }
 
-    pub fn exclude<S: AsRef<str>>(&self, names: &[S]) -> DaftResult<Self> {
-        let mut fields = IndexMap::new();
-        let names = names.iter().map(|s| s.as_ref()).collect::<HashSet<&str>>();
-        for (name, field) in &self.fields {
-            if !names.contains(&name.as_str()) {
-                fields.insert(name.clone(), field.clone());
-            }
-        }
+    pub fn get_field_at_index(&self, index: usize) -> DaftResult<&Field> {
+        self.fields
+            .get(index)
+            .ok_or(DaftError::FieldNotFound(format!(
+                "Attempted to access field at out-of-bounds index {} in schema: {:?}",
+                index, self.fields
+            )))
+    }
 
-        Ok(Self { fields })
+    pub fn fields(&self) -> &[Field] {
+        &self.fields
     }
 
-    pub fn empty() -> Self {
-        Self {
-            fields: indexmap::IndexMap::new(),
-        }
+    pub fn field_names(&self) -> impl Iterator<Item = &str> {
+        self.fields.iter().map(|f| f.name.as_str())
+    }
+
+    #[deprecated(since = "TBD", note = "name-referenced columns")]
+    pub fn exclude<S: AsRef<str>>(&self, names: &[S]) -> Self {
+        let names = names.iter().map(|s| s.as_ref()).collect::<HashSet<&str>>();
+        let fields = self
+            .fields
+            .iter()
+            .filter(|field| !names.contains(field.name.as_str()))
+            .cloned()
+            .collect::<Vec<_>>();
+
+        Self::new(fields)
     }
 
+    #[deprecated(since = "TBD", note = "name-referenced columns")]
     pub fn get_field(&self, name: &str) -> DaftResult<&Field> {
-        match self.fields.get(name) {
-            None => Err(DaftError::FieldNotFound(format!(
+        if let Some(indices) = self.name_to_indices.get(name) {
+            if let [idx] = indices.as_slice() {
+                Ok(&self.fields[*idx])
+            } else {
+                Err(DaftError::FieldNotFound(format!(
+                    "Column name \"{}\" is ambiguous in schema: {:?}",
+                    name, self.fields
+                )))
```

Comment:
Ok great there were two primary functional things here (1) which data structure(s) you went for which look good and (2) how you handle ambiguous errors. Do you think it's worth adding an FieldNameAmbiguous error or keeping with FieldNotFound? The only thing I really care for here is that we error.

Additionally, it may be nice to have the ability get a field case-insensitively 😉

Reply:
Added an error variant for ambiguous reference! 

Field case insensitivity would be useful but I'm going to leave that for another PR